src/SBitSet.jl

```diff
@@ -0,0 +1,284 @@
+
+@inline function count_ones(a::UBitSet{N}) where N 
+    s=0
+    @inbounds for i=1:N 
+        s+= count_ones(a.chunks[i])
+    end
+    s
+end
+
+## Iteration Protocol
+_blsr(x) = x & (x-Int64(1))
+length(a::UBitSet) = count_ones(a)
+eltype(::UBitSet) = Int
+
+@inline function iterate(a::UBitSet{N}) where N
+    N > 0 || return nothing
+    return iterate(a, (1, @inbounds a.chunks[1]))
+end
+
+@inline function iterate(a::UBitSet{N}, s) where N
+    chunks = a.chunks
+    i1, c = s
+    while c==0
+        i1 % UInt >= N % UInt && return nothing
+        i1 += 1
+        @inbounds c = chunks[i1]
+    end
+    tz = trailing_zeros(c) + 1
+    c = _blsr(c)
+    return ((i1-1)<<6 + tz, (i1, c))
+end
+
+## Bitwise operations
+@inline Base.:&(L::UBitSet{N}, R::UBitSet{N}) where N =  SBitSet(ntuple(i->(L.chunks[i] & R.chunks[i]), Val{N}()))
+@inline Base.:|(L::UBitSet{N}, R::UBitSet{N}) where N =  SBitSet(ntuple(i->(L.chunks[i] | R.chunks[i]), Val{N}()))
+@inline xor(L::UBitSet{N}, R::UBitSet{N}) where N =  SBitSet(ntuple(i->xor(L.chunks[i], R.chunks[i]), Val{N}()))
+@inline ~(a::UBitSet{N}) where N = SBitSet(ntuple(i->~a.chunks[i], Val{N}()))
+
+## Broadcast support
+broadcastable(a::UBitSet{N}) where N = Ref(SBitSet{N}(a.chunks))
+
+@inline function materialize!(dst::MBitSet{N}, src::Base.Broadcast.Broadcasted{A, B, typeof(Base.identity),Tuple{Base.RefValue{SBitSet{N}}}} ) where {N,A,B}
+    dst.chunks = src.args[1][].chunks
+    dst
+end
+
+@inline function setindex!(dst::MBitSet{N}, src::UBitSet{N}) where N
+    dst.chunks = src.chunks
+    src
+end
+@inline function getindex(a::MBitSet{N}) where N
+    return SBitSet(a)
+end
+
+## Bounds checking
+@propagate_inbounds function _checkbounds(::UBitSet{N}, idx) where N
+    @boundscheck begin
+    0 < idx <= 64*N || throw(BoundsError(SBitSet{N}, idx))
+    end
+end
+
+@propagate_inbounds function _checkbounds(::Type{Bool}, ::UBitSet{N}, idx) where N
+    @boundscheck begin
+    0 < idx <= 64*N || return false
+    end
+    true
+end
+
+## Individual bit access
+@propagate_inbounds function in(idx, a::UBitSet{N}) where N
+    i1,i2 = Base.get_chunks_id(idx)
+    _checkbounds(Bool, a, idx) || return false
+    @inbounds u = a.chunks[i1]
+    return iszero(u & (1<< (i2 & 63)))
+end
+
+@propagate_inbounds function getindex(a::UBitSet{N}, idx) where N
+    i1,i2 = Base.get_chunks_id(idx)
+    _checkbounds(a, idx)
+    @inbounds u = a.chunks[i1]
+    return iszero(u & (1<< (i2 & 63)))
+end
+
+@propagate_inbounds function setindex!(a::MBitSet{N}, b, idx) where N
+    i1,i2 = Base.get_chunks_id(idx)
+    _checkbounds(a, idx)
+    @inbounds u = a.chunks[i1]
+    nu = ifelse(b, u | (1<< (i2 & 63)), u & ~(1<< (i2 & 63)) )
+    ptr = convert(Ptr{UInt64}, pointer_from_objref(a))
+    GC.@preserve a unsafe_store!(ptr, nu, i1)
+    return b
+end
+
+@propagate_inbounds function Base.setindex(a::SBitSet{N}, b, idx) where N
+    m = MBitSet(a)
+    m[idx] = b
+    return SBitSet(m)
+end
```

Comment:
`SArray` uses an `ifelse` construction instead of pointer arithmetic. LLVM is capable of removing the allocation from the MBitSet, though.

Is this approach ok or should I adopt the `SArray` solution?

Reply:
Since the compiler is getting better, I would go the other way - the `SArray` implementation could quite probably be simplified.

Reply:
I think it depends. Hell, even if I were writing this in assembly I wouldn't know whether to prefer `ifelse` or `pointerset`: `ifelse` is probably faster if the SArray fits into a register. On the other hand, the fact that `pointerset`-style works allocation-free is a somewhat recent development in the compiler, so it is no wonder that StaticArrays uses `ifelse`.

I guess I should run some benchmarks...